flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/hybrid/tiered/tier/disk/DiskCacheManager.java

```diff
@@ -0,0 +1,174 @@
+
+import org.apache.flink.api.java.tuple.Tuple2;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.common.TieredStoragePartitionId;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.file.PartitionFileWriter;
+import org.apache.flink.runtime.io.network.partition.hybrid.tiered.storage.TieredStorageMemoryManager;
+import org.apache.flink.util.concurrent.FutureUtils;
+
+import java.nio.ByteBuffer;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * The {@link DiskCacheManager} is responsible for managing cached buffers before flushing to files.
+ */
+class DiskCacheManager {
+
+    private final TieredStoragePartitionId partitionId;
+
+    private final int numSubpartitions;
+
+    private final PartitionFileWriter partitionFileWriter;
+
+    private final SubpartitionDiskCacheManager[] subpartitionCacheManagers;
+
+    /** Whether the current flush process has completed. */
+    private CompletableFuture<Void> hasFlushCompleted;
+
+    DiskCacheManager(
+            TieredStoragePartitionId partitionId,
+            int numSubpartitions,
+            TieredStorageMemoryManager storageMemoryManager,
+            PartitionFileWriter partitionFileWriter) {
+        this.partitionId = partitionId;
+        this.numSubpartitions = numSubpartitions;
+        this.partitionFileWriter = partitionFileWriter;
+        this.subpartitionCacheManagers = new SubpartitionDiskCacheManager[numSubpartitions];
+        this.hasFlushCompleted = FutureUtils.completedVoidFuture();
+
+        for (int subpartitionId = 0; subpartitionId < numSubpartitions; ++subpartitionId) {
+            subpartitionCacheManagers[subpartitionId] = new SubpartitionDiskCacheManager();
+        }
+        storageMemoryManager.listenBufferReclaimRequest(this::notifyFlushCachedBuffers);
+    }
+
+    // ------------------------------------------------------------------------
+    //  Called by DiskTierProducerAgent
+    // ------------------------------------------------------------------------
+
+    /**
+     * Append buffer to {@link DiskCacheManager}.
+     *
+     * @param buffer to be managed by this class.
+     * @param subpartitionId the subpartition of this record.
+     */
+    void append(Buffer buffer, int subpartitionId) {
+        subpartitionCacheManagers[subpartitionId].append(buffer);
+    }
+
+    /**
+     * Append the end-of-segment event to {@link DiskCacheManager}, which indicates the segment has
+     * finished.
+     *
+     * @param record the end-of-segment event
+     * @param subpartitionId target subpartition of this record.
+     * @param dataType the type of this record. In other words, is it data or event.
+     */
+    void appendEndOfSegmentEvent(ByteBuffer record, int subpartitionId, Buffer.DataType dataType) {
```

Comment:
`Buffer.DataType` dataType seems can be removed safety.

Reply:
This should not be removed, because the previous comments show that the concept of end-of-segment need not be aware both in `SubpartitionDiskCacheManager` and `DiskProducerAgent`. I also think this makes sense. 

So the `Buffer.DataType` is to avoid composing the end-of-segment to `SubpartitionDiskCacheManager`.

Reply:
What confuses me is: this method called `appendEndOfSegmentEvent`, Isn't this implying that it must be an event? Is it possible that it will be `Data` in the future? 

If we have a method called `emitEvent (Buffer event)` in a class `A`, I don't think this can be considered as `A` exactly aware the specific event.